File: clients/ts/src/hopKind.ts
import { assertNever } from "./assertNever.js";
import { FindRouteError } from "./error.js";

/**
 * A venue the menu may name; the value is the byte the wire carries.
 *
 * The program dispatches on a wider numbering; only these ten are accepted in a `find_route` menu,
 * and the others cannot be spelled here: the type is the union of these ten literals.
 */
export const HopKind = {
  RaydiumAmmV4: 0,
  Whirlpool: 1,
  RaydiumClmm: 2,
  RaydiumCpmm: 3,
  MeteoraDlmmSwap: 4,
  MeteoraDlmmSwap2: 5,
  MeteoraDammV2: 6,
  PumpSwapSell: 7,
  PumpSwapBuy: 8,
  MeteoraDammV1: 9,
} as const;

export type HopKind = (typeof HopKind)[keyof typeof HopKind];

export type HopKindName = keyof typeof HopKind;

/** Every kind, in wire order. */
export const ALL_HOP_KINDS: ReadonlyArray<HopKind> = [
  HopKind.RaydiumAmmV4,
  HopKind.Whirlpool,
  HopKind.RaydiumClmm,
  HopKind.RaydiumCpmm,
  HopKind.MeteoraDlmmSwap,
  HopKind.MeteoraDlmmSwap2,
  HopKind.MeteoraDammV2,
  HopKind.PumpSwapSell,
  HopKind.PumpSwapBuy,
  HopKind.MeteoraDammV1,
];

export function isHopKind(value: number): value is HopKind {
  return ALL_HOP_KINDS.some((kind) => kind === value);
}

/** The kind a byte names; throws `UnknownHopKind` for any other byte. */
export function hopKindFromByte(raw: number): HopKind {
  if (isHopKind(raw)) {
    return raw;
  }
  throw new FindRouteError({ kind: "UnknownHopKind", raw });
}

/** The manifest's name for a kind. */
export function hopKindName(kind: HopKind): HopKindName {
  switch (kind) {
    case HopKind.RaydiumAmmV4:
      return "RaydiumAmmV4";
    case HopKind.Whirlpool:
      return "Whirlpool";
    case HopKind.RaydiumClmm:
      return "RaydiumClmm";
    case HopKind.RaydiumCpmm:
      return "RaydiumCpmm";
    case HopKind.MeteoraDlmmSwap:
      return "MeteoraDlmmSwap";
    case HopKind.MeteoraDlmmSwap2:
      return "MeteoraDlmmSwap2";
    case HopKind.MeteoraDammV2:
      return "MeteoraDammV2";
    case HopKind.PumpSwapSell:
      return "PumpSwapSell";
    case HopKind.PumpSwapBuy:
      return "PumpSwapBuy";
    case HopKind.MeteoraDammV1:
      return "MeteoraDammV1";
    default:
      return assertNever(kind);
  }
}

File: clients/rust/turk-router/src/hop_kind.rs
use crate::Error;

/// A venue the menu may name. The discriminant is the byte the wire carries.
///
/// The program dispatches on a wider numbering; only these ten are accepted in a `find_route`
/// menu, and the others cannot be constructed here. Deliberately not `#[non_exhaustive]`: the set
/// is the contract, and a `match` over it should break when a kind is added.
#[repr(u8)]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum HopKind {
    /// Raydium AMM v4.
    RaydiumAmmV4 = 0,
    /// Orca Whirlpool.
    Whirlpool = 1,
    /// Raydium concentrated liquidity (CLMM).
    RaydiumClmm = 2,
    /// Raydium constant-product (CPMM).
    RaydiumCpmm = 3,
    /// Meteora DLMM, the `swap` instruction.
    MeteoraDlmmSwap = 4,
    /// Meteora DLMM, the `swap2` instruction.
    MeteoraDlmmSwap2 = 5,
    /// Meteora DAMM v2.
    MeteoraDammV2 = 6,
    /// PumpSwap, selling the base token.
    PumpSwapSell = 7,
    /// PumpSwap, buying the base token.
    PumpSwapBuy = 8,
    /// Meteora DAMM v1.
    MeteoraDammV1 = 9,
}

impl HopKind {
    /// Every kind, in wire order.
    pub const ALL: [HopKind; 10] = [
        HopKind::RaydiumAmmV4,
        HopKind::Whirlpool,
        HopKind::RaydiumClmm,
        HopKind::RaydiumCpmm,
        HopKind::MeteoraDlmmSwap,
        HopKind::MeteoraDlmmSwap2,
        HopKind::MeteoraDammV2,
        HopKind::PumpSwapSell,
        HopKind::PumpSwapBuy,
        HopKind::MeteoraDammV1,
    ];

    /// The byte the menu entry carries.
    #[must_use]
    pub const fn discriminant(self) -> u8 {
        self as u8
    }
}

impl TryFrom<u8> for HopKind {
    type Error = Error;

    fn try_from(raw: u8) -> Result<Self, Error> {
        HopKind::ALL
            .into_iter()
            .find(|kind| kind.discriminant() == raw)
            .ok_or(Error::UnknownHopKind { raw })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn discriminants_are_dense_from_zero() {
        for (index, kind) in HopKind::ALL.iter().enumerate() {
            assert_eq!(usize::from(kind.discriminant()), index);
            assert_eq!(HopKind::try_from(kind.discriminant()), Ok(*kind));
        }
    }

    #[test]
    fn bytes_past_the_menu_set_are_refused() {
        for raw in 10u8..=u8::MAX {
            assert_eq!(HopKind::try_from(raw), Err(Error::UnknownHopKind { raw }));
        }
    }
}
